client/hijack.go

```diff
@@ -140,23 +167,37 @@ func tlsDialWithDialer(dialer *net.Dialer, network, addr string, config *tls.Con
 	return &tlsClientCon{conn, rawConn}, nil
 }
 
-func dial(proto, addr string, tlsConfig *tls.Config) (net.Conn, error) {
+func dial(ctx context.Context, proto, addr string, tlsConfig *tls.Config) (net.Conn, error) {
+	// 0 = no timeout
+	var timeout time.Duration
+
+	deadline, ok := ctx.Deadline()
+	if ok {
+		timeout = time.Until(deadline)
+	}
+
 	if tlsConfig != nil && proto != "unix" && proto != "npipe" {
 		// Notice this isn't Go standard's tls.Dial function
-		return tlsDial(proto, addr, tlsConfig)
+		return tlsDialWithDialer(&net.Dialer{Timeout: timeout}, proto, addr, tlsConfig)
 	}
 	if proto == "npipe" {
-		return sockets.DialPipe(addr, 32*time.Second)
+		if timeout == 0 {
+			// Why 32? See issue 8035
```

Comment:
Oh, LOL, I recall that discussion :joy: https://github.com/moby/moby/pull/8035#issuecomment-55645549